Initial creation of Wiki class to hold data from OSF API.

Trumbore opened the pull request:
Elliot,

Here is an initial implementation for the Wiki class.  I used the File class as a model and copied code from there, adding missing elements as needed.  You don't have to accept this as a pull request, it's mainly so we can discuss my work and what comes next.  Here are some notes I made as I worked on the code.

• Looking at the File class, it seems like it is for both a folder and a file.
    o       Folder has relationships: files, node.  File has relationships: node, comments, versions.
• In File, Node is stored as an attribute - is that because there can be only one?  By that logic, I made User an attribute.
• I left out the pageLinks found in File, but wasn't sure what it did (it did not appear in the OSF JSON).
• File does not include some of its attributes: last_touched, current_user_can_comment, guid, checkout, tags.
    o       I left current_user_can_comment out of Wiki as well, but added content_type, which might not be needed.
• Much as "hashes" was found in File's "extra" section, I included "version" from Wiki's extra section.
• "type" was not included in File, so I left it out of Wiki.
• Is File's "links" for the top level links, or the one inside "data"?  Wiki only has the one inside "data".
• Import statements were just copied from File.  I will clean these up after annotations are complete.
• I kept the property ordering that was found in File.  Would it be OK to reorder them as they appear in Wiki OSF JSON?

Ben

Trumbore committed:
Merge remote-tracking branch 'refs/remotes/DataConservancy/master'

Trumbore committed:
Merge remote-tracking branch 'refs/remotes/DataConservancy/master'

Trumbore committed:
Initial creation of Wiki class to hold data from OSF API.

emetsger commented:
Hey @Trumbore I'll comment in more detail, but it would be nice if the 3fcdb39 commit was rebased against the tip of master, so that the two merge commits 94a6aa9 and 292de1b did not show in the history of your branch.  Ideally the commit history for this branch would not have any merge commits, but if that's too much trouble, no worries, I can make it work!

emetsger commented on src/main/java/org/dataconservancy/cos/osf/client/model/Wiki.java:
```diff
@@ -0,0 +1,214 @@
+ */
+//@Type("wiki")
+//@JsonIgnoreProperties(ignoreUnknown = true)
+//@OwlIndividual(OwlClasses.OSF_WIKI)
```
In order to more easily perform the review, could you remove all of the comments?  They clutter up the code, and aren't necessary at this time.  We can add them back in as development progresses!

Trumbore replied:
Done.  I'm going to leave most of the import statements, though, as they will be needed when we start adding annotations.

emetsger replied:
+1

emetsger commented on src/main/java/org/dataconservancy/cos/osf/client/model/Wiki.java:
```diff
@@ -0,0 +1,214 @@
+
+    /**version of wiki*/
+    //@OwlProperty(OwlProperties.OSF_HAS_SIZE)
+    private Integer version;
```
We may need to have a field for `extra`, which would contain the `version` property, rather than have `version` on its own.

Trumbore replied:
I followed the way it was done in File, which includes hashes directly without having the "extra" from the JSON.  If I add extra to Wiki, what would it be, a Map<string, ?>?

emetsger replied:
Yes, IIRC a map with `String` keys and the unknown type as the value (`<?>`)

emetsger commented on src/main/java/org/dataconservancy/cos/osf/client/model/Wiki.java:
```diff
@@ -0,0 +1,214 @@
+
+    /**size of wiki in bytes, null for folders*/
+    //@OwlProperty(OwlProperties.OSF_HAS_SIZE)
+    private Integer size;
```
Because `-1` is a more sane default than `null`, let's use the primitive `int` rather than the boxed `Integer`.

Trumbore replied:
I made this change for both size and version.  If you only meant it for size, let me know.

emetsger replied:
+1

Trumbore commented:
I'm afraid I don't think I can rebase.  I tried it and ran got the following:

C:\Projects\IRD\IdeaProjects\osf-packaging [master ↑]> git rebase upstream/master
First, rewinding head to replay your work on top of it...
Applying: Delete files with bad names
Using index info to reconstruct a base tree...
A       osf-model/src/test/resources/json/OsfPackageGraphTest/testCreatePackageGraph/nodes/3e7rd/comments/?filter[target]=jmfd3index-01.json
[...then the same for the other 25 files with bad names...]
Falling back to patching base and 3-way merge...
No changes -- Patch already applied.

The output text looks like what I see in online explanations of rebasing, but I know that adding those files back must have failed, and am worried that means that Git doesn't think it's up to date.

emetsger commented:
Ok, yeah you probably can't rebase because of the issues with the file names.  Don't worry about that.  I'll handle it when it is time to merge

emetsger commented:
The next step would be to write a unit test, demonstrating that the `OsfService` can parse wiki information presented by the JSON API into the `Wiki` class.

Trumbore committed:
Respond to review comments for initial Wiki class

Trumbore committed:
Merge remote-tracking branch 'upstream/master'

Trumbore committed:
Create a unit test for parsing OSF Wiki data into the Java class.

Works for a lot of the relationships and attributes, but not for
comments, the extra attribute, and three attributes whose names include
underscores.

Trumbore committed:
Try again to commit changes to two files (one new)

Trumbore commented:
Hi Elliot,

After more trial and error, I got the majority of the Wiki unit test to work.  Here are some notes:
• I couldn't figure out what to do with the "comments" relationship.  I kept getting the error where it couldn't map the URL to a local file.  Possibly because of Windows path issues?  Tried without the filter at the end of the URL, but no joy.  I included JSON files where I think they should go, though the leaf comment file contains error JSON because I could not create a wiki comment through the OSF site.
• Three simple attributes whose tests are commented out in the code are always returned as null after parsing (materialized_path, date_modified, content_type).  Are these attributes skipped in other tests, or is there a way to enable their parsing?  I notice that all three names include underscores, and none of the other attribute names do that, so maybe this is another Windows thing.
• A fourth attribute ("extra") is always returned as null, and it would contain the "version" info we would like to access.  Is there a way to enable the parsing of this attribute?  Until it is parsed, the getVersion and setVersion methods are just placeholders that operate on the whole "extra" Map.
• I included the file we discussed for the wiki "content", but testing that will come as part of the conversion to RDF, right?

I look forward to hearing your thoughts on these remaining issues.  Hanh asked me at today's standup what my next step would be after this test is working, and I deferred to you.  Just a heads up that it might come up in Friday's standup.

```
Ben 
```

emetsger commented on src/test/java/org/dataconservancy/cos/osf/client/model/WikiTest.java:
```diff
@@ -0,0 +1,75 @@
+        assertTrue(pjnbm.getUser().endsWith("3rty2/"));
+        // TODO - Can something be included for "comments"?
+
+        assertEquals(pjnbm.getKind(), "file");
```
The expectations for the _assert*_ methods are backwards: the expected value should go first, followed by the actual value.

Trumbore replied:
I have made these changes

emetsger commented on src/main/java/org/dataconservancy/cos/osf/client/model/Wiki.java:
```diff
@@ -141,23 +110,23 @@ public void setName(String name) {
     }
 
-    public void setDate_modified(String date_modified) {
+    public void setDateModified(String date_modified) {
```
These methods should be named after the name of the field: so the original name, `setDate_modified` is correct.  That is why these values are not being set: the method name is incorrect.

emetsger commented on src/main/java/org/dataconservancy/cos/osf/client/model/Wiki.java:
```diff
@@ -141,23 +110,23 @@ public void setName(String name) {
     }
 
-    public void setMaterialized_path(String materialized_path) {
+    public void setMaterializedPath(String materialized_path) {
```
These methods should be named after the name of the field: so the original name, setMaterialized_path is correct. That is why these values are not being set: the method name is incorrect.

emetsger commented on src/main/java/org/dataconservancy/cos/osf/client/model/Wiki.java:
```diff
@@ -0,0 +1,190 @@
+        return content_type;
+    }
+
+    public void setContentType(String content_type) {
```
Same issue here with setter method name.  Correct the name to be `setContent_type` and the value ought to be set properly.

emetsger commented on src/main/java/org/dataconservancy/cos/osf/client/model/Wiki.java:
```diff
@@ -0,0 +1,190 @@
+        this.links = links;
+    }
+
+    public Map<String, ?> getVersion() {
```
This needs to be named after the field `extra`, like the others.  So `setExtra` and `getExtra`

emetsger commented on src/test/java/org/dataconservancy/cos/osf/client/model/WikiTest.java:
```diff
@@ -0,0 +1,75 @@
+        //assertEquals(pjnbm.getMaterializedPath(), "/pjnbm");
+        assertEquals(pjnbm.getSize(), 916);
+        assertEquals(pjnbm.getId(), "pjnbm");
+        // TODO - What to do about extra, which contains version?
```
You would retrieve the Map (`getExtra`) and insure that the `version` key has the expected value.

Trumbore replied:
This, along with the three answers below, makes more clear to me how the association between JSON properties and Java fields are magically performed!  Yes, all four changes make things work.

emetsger commented on src/test/resources/json/WikiTest/testWikiMapping/registrations/ng9em/wikis/index.json:
```diff
@@ -0,0 +1,54 @@
+{
```
I would say if this file is not being used by the test, then we should remove it from this PR.  It's hard to tell if it is necessary.

I would keep the comments file, because we'll want to test that in the future.  (I think I have a PR that will help with that).

Trumbore replied:
This is actually the list of wikis that is loaded at the start of the test, after which I get the first wiki and look at it.  So, the file is needed with the test as is.  I could alter the test to just look at one wiki, if you'd like.

emetsger reviewed (changes requested):
Hi Ben, looks pretty good.  I left some comments which need to be corrected, which will ultimately lead  to a fully-passing test!

Trumbore committed:
Respond to review comments for Wiki JSON parsing
